068-access-policies-in-jwt.md

```diff
@@ -0,0 +1,183 @@
+    "cluster_x:topic_1:read",
+    ":topic2:write"
+  ]
+}
+```
+
+and for a string:
+```json
+{
+  "acls": "cluster_x:topic_1:read,topic2:write"
+}
+```
+
+The field name is configurable with default value `acls`.
+
+The new authorizer will be implemented as a new class, similarly to
+[KeycloakAuthorizer](https://github.com/strimzi/strimzi-kafka-oauth/blob/229daee85b096804d16e2904c8c0f1add599cc99/oauth-keycloak-authorizer/src/main/java/io/strimzi/kafka/oauth/server/authorizer/KeycloakAuthorizer.java),
+making use of the [KeycloakRBACAuthorizer](https://github.com/strimzi/strimzi-kafka-oauth/blob/229daee85b096804d16e2904c8c0f1add599cc99/oauth-keycloak-authorizer/src/main/java/io/strimzi/kafka/oauth/server/authorizer/KeycloakRBACAuthorizer.java).
+This is in order to support both KRaft and Zookeeper.
+
+```java
+public class JWTAuthorizer implements ClusterMetadataAuthorizer {
+    private StandardAuthorizer delegate;
+    private KeycloakRBACAuthorizer singleton;
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+        // ...
+    }
+
+    @Override public List<AuthorizationResult> authorize(AuthorizableRequestContext requestContext, List<Action> actions) {
+        KafkaPrincipal principal = requestContext.principal();
+        if (!(principal instanceof OAuthKafkaPrincipal)) {
+          // simple ACL delegation
+        }
+        BearerTokenWithPayload token = ((OAuthKafkaPrincipal) principal).getJwt();
+        // check token validity
+        BearerTokenWithPayload jwt = principal.getJwt();
+        ObjectNode payload = jwt.getClaimsJSON();
+        // get the acls from the payload
+        // for each access request
+        //   add to the result whether the topic or cluster claims allow the access request
+        // return the result
+    }
+    
+    // ...
+}
+```
+
+Operations are described in the [Kafka docs](https://kafka.apache.org/documentation/#operations_in_kafka).
+Allowing an operation will follow the same logic as in [AclEntry.supportedOperations](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/security/authorizer/AclEntry.scala#L99).
+
+The above code will be supplied with unit tests and documentation explaining how to set up Strimzi with
+the new authorizer using Azure Active Directory B2C as an example.
+The setup will be similar for other identity providers, with the main difference being the configuration of the identity provider itself and the `acls` field.
+
+## Scalability
+
+This solution is intended for small to medium-sized clients.
+In practice the permissions that are granted to a client are usually at most a few dozen.
+In most cases in microservices, even less than ten as the client is usually designed to perform a single task,
+i.e., read from a single topic (or a few) and write to a single topic.
+If a client has hundreds of permissions it is likely that architecture is flawed or
+the client performs too many tasks, which is again a sign of a flawed architecture.
```

Comment:
I don't think this is true. It is from my experience quite common to consume from many topics. You have also Kafka Streams applications that consume and produce many topics if they do some complicated stateful operations. So I do not think that is a flawed architecture.

Reply:
I have only once worked with consuming data from more than 20 topics, and it was not a nice experience.
Calling it flawed was my bad, I will remove the comment.

Reply:
To be honest, I think you should maybe remove the whole subjective section on how you think Kafka should be used and instead focus on the hard data of the actual scalability and limitations. I.e. things such as:
* With how many permissions is it expected to work
* With how many permissions does it break
* What is the performance and memory impact for each

Once the limits are clearly listed, people can decide on their own of how useful they are.

For me personally, I don't think 100s of permissions would be common. But 10s of them are definitely very common.

Reply:
I'll rework it, leaving only the numbers.